osu.Game/Overlays/DirectOverlay.cs

```diff
@@ -225,28 +249,122 @@ protected override void PopIn()
             base.PopIn();
 
             // Queries are allowed to be run only on the first pop-in
-            if (getSetsRequest == null)
+            if (!beatmapSetPager.IsFetching)
                 queueUpdateSearch();
         }
 
-        private SearchBeatmapSetsRequest getSetsRequest;
-
         private readonly Bindable<string> currentQuery = new Bindable<string>(string.Empty);
 
         private ScheduledDelegate queryChangedDebounce;
+        private ScheduledDelegate addPageDebounce;
         private PreviewTrackManager previewTrackManager;
+        private BeatmapSetPager beatmapSetPager;
+
+        private class BeatmapSetPager
+        {
+            public event PageFetchHandler PageFetch;
+
+            private readonly RulesetStore rulesets;
+
+            private SearchBeatmapSetsRequest getSetsRequest;
+
+            private int currentPage = 1;
+
+            public bool IsLastPageFetched { get; private set; } = false;
+            public bool IsFetching => getSetsRequest != null;
+
+            public BeatmapSetPager(RulesetStore rulesets)
+            {
+                this.rulesets = rulesets;
+            }
+
+            public SearchBeatmapSetsRequest FetchNextPage(string query, RulesetInfo ruleset, BeatmapSearchCategory searchCategory = BeatmapSearchCategory.Any, DirectSortCriteria sortCriteria = DirectSortCriteria.Ranked, SortDirection direction = SortDirection.Descending)
+            {
+                if (getSetsRequest != null)
+                    return null;
+
+                getSetsRequest = new SearchBeatmapSetsRequest(
+                    query,
+                    ruleset,
+                    currentPage,
+                    searchCategory,
+                    sortCriteria,
+                    direction);
+
+                getSetsRequest.Success += response =>
+                {
+                    Task.Run(() =>
+                    {
+                        var sets = response.BeatmapSets.Select(r => r.ToBeatmapSet(rulesets)).ToList();
+
+                        if (sets.Count <= 0) IsLastPageFetched = true;
+
+                        PageFetch?.Invoke(currentPage, sets);
+                        
+                        getSetsRequest = null;
+                        currentPage++;
+                    });
```

Comment:
Looks fragile due to the fire-and-forget `Task.Run`. What happens if any of the operations fail? I'm worried they'll fail silently.

Why's this in a task in the first place? I don't think any of this is inherently costly.

Reply:
It was just a task originally, so I just thought to maintain the status quo. I'll remove it.